pool_manager.go

```diff
@@ -0,0 +1,104 @@
+package apns2
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+	"unsafe"
+
+	"golang.org/x/net/http2"
+	"golang.org/x/net/idna"
+)
+
+type poolManager struct {
+	connPool http2.ClientConnPool
+	ctx      context.Context
+	poolMu   *sync.Mutex
+	u        *url.URL
+}
+
+// directly copied from source
+func authorityAddr(scheme string, authority string) (addr string) {
+	host, port, err := net.SplitHostPort(authority)
+	if err != nil {
+		port = "443"
+		if scheme == "http" {
+			port = "80"
+		}
+		host = authority
+	}
+	if a, err := idna.ToASCII(host); err == nil {
+		host = a
+	}
+	if strings.HasPrefix(host, "[") && strings.HasSuffix(host, "]") {
+		return host + ":" + port
+	}
+	return net.JoinHostPort(host, port)
+}
+
+func newPoolManager(transport *http2.Transport, environment string) (*poolManager, error) {
+	transport.CloseIdleConnections()
+	rf := reflect.Indirect(reflect.ValueOf(transport)).FieldByName("connPoolOrDef")
+	connPool := *(*http2.ClientConnPool)(unsafe.Pointer(rf.UnsafeAddr()))
+	rf = reflect.Indirect(reflect.ValueOf(connPool)).FieldByName("mu")
+	poolMu := (*sync.Mutex)(unsafe.Pointer(rf.UnsafeAddr()))
+	u, err := url.Parse(environment)
+	if err != nil {
+		return nil, err
+	}
+	return &poolManager{
+		connPool: connPool,
+		u:        u,
+		poolMu:   poolMu,
+		ctx:      context.Background(),
+	}, nil
+}
+
+func (pm *poolManager) addNewConn() error {
+	rff := reflect.Indirect(reflect.ValueOf(pm.connPool)).FieldByName("conns")
+	pm.poolMu.Lock()
+	internalConns := *(*map[string][]*http2.ClientConn)(unsafe.Pointer(rff.UnsafeAddr()))
+	for _, conns := range internalConns {
+		for _, conn := range conns {
+			rv := reflect.Indirect(reflect.ValueOf(conn))
+			rf := rv.FieldByName("mu")
+			mu := (*sync.Mutex)(unsafe.Pointer(rf.UnsafeAddr()))
+			rf = rv.FieldByName("closed")
+			mu.Lock()
+			rf = reflect.Indirect(reflect.NewAt(rf.Type(), unsafe.Pointer(rf.UnsafeAddr())))
+			rf.SetBool(true)
+			mu.Unlock()
+			defer func() {
```

Comment:
I think `mu` and `rf` should be passed as parameters here to avoid issue like here: https://play.golang.org/p/KKaK0776AK